tsfc/coffee_mode.py

```diff
@@ -221,28 +220,28 @@ def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
     """
     if not optimal_atomics:
         return monomial_sum
-    if len(monomial_sum.ordering) < 2:
+    if len(monomial_sum) < 2:
         return monomial_sum
     new_monomial_sum = MonomialSum()
     # Group monomials with respect to each optimal atomic
     factor_group = OrderedDict()
     for monomial in monomial_sum:
-        for sum_indices, oa in optimal_atomics:
-            if frozenset(monomial.sum_indices) == frozenset(sum_indices) and oa in monomial.atomics:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
                 # Add monomial to the list of corresponding optimal atomic
-                factor_group.setdefault((sum_indices, oa), []).append(monomial)
+                factor_group.setdefault(oa, []).append(monomial)
                 break
         else:
             # Add monomials that do no have argument factors to new MonomialSum
-            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+            new_monomial_sum.add(*monomial)
```

Comment:
There could be more than one atomics in the intersection, so I write it as a for loop with break, to group the monomials with respect to the first optimal atomic they have.